modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/rel/MinusNode.java

```diff
@@ -0,0 +1,341 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.query.calcite.exec.rel;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.ignite.internal.processors.query.calcite.exec.ExecutionContext;
+import org.apache.ignite.internal.processors.query.calcite.exec.RowHandler;
+import org.apache.ignite.internal.processors.query.calcite.exec.RowHandler.RowFactory;
+import org.apache.ignite.internal.processors.query.calcite.exec.exp.agg.AggregateType;
+import org.apache.ignite.internal.processors.query.calcite.exec.exp.agg.GroupKey;
+import org.apache.ignite.internal.util.typedef.F;
+
+/**
+ * Execution node for MINUS (EXCEPT) operator.
+ */
+public class MinusNode<Row> extends AbstractNode<Row> {
```

Comment:
I may not really understand how the EXCEPT is supposed to work... So let's find out it first to sure we are on the same page. My understanding is matches the [Postgres definition](https://www.postgresql.org/docs/7.4/sql-select.html) (sorry, but there is no anchor for exact place).

Current implementation of the execution node says that only leftmost query defines a top bound of row count, every next stage of EXCEPT could only shorten the resultset. Hence a conclusion: we could drop following stages as soon as current stage returns empty resultset. It's true that we have to complete gathering rows for current stage, but I see no reason to proceed the work a result of which we are no more interested in.


> Did you mean [all=true] (not DISTINCT)?

No, I meant exactly what I said. Frankly I have no clue how a query with EXCEPT ALL could be easily rewritten with ANTI JOIN since it requires a counting rows on both sides. On the other hand a query like `SELECT f1 FROM t1 EXCEPT SELECT f1 FROM t2` could be simply rewritten to `SELECT DISTINCT f1 FROM t1 ANTI JOIN t2 USING (f1)`.


> Calcite uses about the same approach

For some (unclear to me) reasons Calcite doesn't like SEMI and ANTI joins. And we already faced a [problem](https://issues.apache.org/jira/browse/IGNITE-14277) because of this

Reply:
> My understanding is matches the Postgres definition

Yes, the current implementation matches the Postgres definition too. I got the idea with this optimization, I will implement it. But as I said before, is not always possible, for example, we still need to process all sources on the MAP phase.

> On the other hand a query like `SELECT f1 FROM t1 EXCEPT SELECT f1 FROM t2` could be simply rewritten to `SELECT DISTINCT f1 FROM t1 ANTI JOIN t2 USING (f1)`

I'm not sure it's more optimal than the current approach. As far as I understand there are some limitations where this approach is applicable: tables `t1` and `t2` should be colocated, there should be an index on `t2` for `f1` columns, `t1` should be sorted by `f1` or streaming mode will be not possible with DISTINCT, etc.
I propose to start with the current implementation (as working for all cases and optimal for most cases) and make optimization with anti-join later if we really need it.

Reply:
> we still need to process all sources on the MAP phase

`select_1 EXCEPT select_2 EXCEPT select_3 EXCEPT select_n` should be evaluated as `((select_1 EXCEPT select_2) EXCEPT select_3) EXCEPT select_n`. Could you please explain why we need to process rows from `select_n` if results of `(...) EXCEPT select_3` is empty? It's unclear to me right now.


> I propose to start with the current implementation

works for me. It's better to benchmark both approaches and compare results to decide which one is better.

Reply:
> Could you please explain why we need to process rows from select_n on MAP phase if results of `(...) EXCEPT select_3` is empty?

For example, we have 'a' EXCEPT 'a' EXCEPT 'b' on node 1, and 'b' EXCEPT 'c' EXCEPT 'c' on node 2. In this case after the first step, we already have an empty set on node 1. If we stop processing other inputs on node 1, on reduce phase we will have the resulting set as 'b'. But this 'b' (from node 2) should be eliminated by the result from node 1.